src/qtism/data/AssessmentTest.php

```diff
@@ -430,17 +432,207 @@ public function isExclusivelyLinear()
     }
+
+    /**
+     * Returns an array with all possible paths for a AssessmentTest.
+     * @param asArray true to return an array, false to return an AssessmentItemRefCollection.
+     * @return array of array of AssessmentItemRef or an array AssessmentItemRefCollection, depending on $asArray.
+     * @throws Exception if branching is recursive of backward.
+     */
+
+    public function getPossiblePaths($asArray)
+    {
+        $paths = array();
+        $items = new AssessmentItemRefCollection($this->getComponentsByClassName("assessmentItemRef")->getArrayCopy());
+        $paths[] = $items;
+        $identifier_list = $items->getKeys(); // list of the identifiers
+        $id_to_index = array_flip($identifier_list); // get the index of the item with its ID, needed to order branches
+
+        // Array associating to each item the possible successor item
+
+        $succs = array();
+
+        // Association of the successor item to the next one
+
+        for ($i = 0; $i < count($items); $i++) {
+            $succs[$identifier_list[$i]] = array();
+
+            if ($i < (count($items) - 1)) {
+                $succs[$identifier_list[$i]][] = $items[$identifier_list[$i + 1]];
+            }
+        }
+
+        // Checking existing branches to add other possible previous items
+
+        foreach ($items as $item) {
+
+            foreach ($item->getComponentsByClassName("branchRule") as $branch) {
+
+                $item_target = $this->getComponentByIdentifier($branch->getTarget());
+
+                if (!in_array($item_target, $succs[$item->getIdentifier()])) {
+                    $target = $this->getComponentByIdentifier($branch->getTarget());
+                    $succs[$item->getIdentifier()][] = $target;
+
+                    // new successor possible => new paths possible
+
+                    foreach ($paths as $path) {
+
+                        // get the index of the current item and of the target item
+
+                        $key_current_item = null;
+                        $key_target_item = null;
+
+                        $pathkeys = $path->getKeys();
+
+                        foreach ($pathkeys as $identifier) {
+
+                            if ($item->getIdentifier() == $identifier) {
+                                $key_current_item = $item->getIdentifier();
+                            }
+
+                            if ($branch->getTarget() == $identifier) {
+                                $key_target_item = $branch->getTarget();
+                            }
+                        }
+
+                        if (($key_current_item != null) and ($key_target_item != null)) {
+
+                            if ($key_current_item == $key_target_item) {
+                                throw new \Exception("Recursive branching is not allowed."); //http://stackoverflow.com/questions/22698093/disable-warning-the-thrown-object-must-be-an-instance-of-the-exception
```

Comment:
The idea is good and this is what we want! However, I would like you to go a little bit further by creating your own specific Exception (e.g. PathException, ... please find the most appropriate name) from where you could extract the most information as possible e.g. (exception code [see $code in \Exception class], this will help the client code (calling your API) to know more about the intrinsic nature of the error, sometimes far above in the call stack. 

**_Note: Same thing with all \Exception instances thrown. Please find the best balance between:**

A single specific exception class, which is more general about path discovery.
Multiple exception classes that are more specific.

(I don't know what is the best yet for this particular case. Please motivate your choice [even if it is in between] by strong arguments)._